addons/sale_purchase/models/sale_order.py

```diff
@@ -174,13 +174,17 @@ def _purchase_increase_ordered_qty(self, new_qty, origin_values):
         """
         for line in self:
             last_purchase_line = self.env['purchase.order.line'].search([('sale_line_id', '=', line.id)], order='create_date DESC', limit=1)
-            if last_purchase_line.state in ['draft', 'sent', 'to approve']:  # update qty for draft PO lines
+            if line.need_update_quantity(last_purchase_line):  # update qty for draft PO lines
                 quantity = line.product_uom._compute_quantity(new_qty, last_purchase_line.product_uom)
                 last_purchase_line.write({'product_qty': quantity})
-            elif last_purchase_line.state in ['purchase', 'done', 'cancel']:  # create new PO, by forcing the quantity as the difference from SO line
+            else:  # create new PO, by forcing the quantity as the difference from SO line
```

Comment:
I think it would be better two create two methods for the two conditions. Because with your change,  it's not possible to skip those two processes and do something else.